pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		wireName := prop.name
+		propertyName := csharpPropertyName(prop)
+		propertyType := qualifiedCSharpPropertyType(prop, "Outputs", false /*wrapInput*/)
+
+		// Workaround the fact that provider inputs come back as strings.
+		if rg.res.IsProvider() && !isPrimitiveKind(prop.typ.kind) {
+			propertyType = "string"
+			if prop.optional() {
+				propertyType += "?"
+			}
+		}
+
+		emitCSharpPropertyDocComment(rg.w, prop, "        ")
+		rg.w.Writefmtln("        [Output(\"%s\")]", wireName)
+		rg.w.Writefmtln("        public Output<%s> %s { get; private set; } = null!;", propertyType, propertyName)
+		rg.w.Writefmtln("")
+	}
+	rg.w.Writefmtln("")
+
+	// Emit the class constructor.
+	argsType := rg.res.argst.name
+
+	var argsDefault string
+	if len(rg.res.reqprops) == 0 {
+		// If the number of required input properties was zero, we can make the args object optional.
+		argsDefault = " = null"
+		argsType += "?"
+	}
+
+	optionsType := "CustomResourceOptions"
+	if rg.res.IsProvider() {
+		optionsType = "ResourceOptions"
+	}
+
+	// Write a comment prior to the constructor.
+	rg.w.Writefmtln("        /// <summary>")
+	rg.w.Writefmtln("        /// Create a %s resource with the given unique name, arguments, and options.", className)
+	rg.w.Writefmtln("        ///")
+	rg.w.Writefmtln("        /// <param name=\"name\">The unique name of the resource</param>")
+	rg.w.Writefmtln("        /// <param name=\"args\">The arguments used to populate this resource's properties</param>")
+	rg.w.Writefmtln("        /// <param name=\"options\">A bag of options that control this resource's behavior</param>")
+	rg.w.Writefmtln("        /// </summary>")
+
+	rg.w.Writefmtln("        public %s(string name, %s args%s, %s? options = null)",
+		className, argsType, argsDefault, optionsType)
+	rg.w.Writefmtln("            : base(\"%s\", name, args, MakeResourceOptions(options, \"\"))", rg.res.info.Tok)
+	rg.w.Writefmtln("        {")
+	rg.w.Writefmtln("        }")
+
+	// Write a private constructor for the use of `Get`.
+	if !rg.res.IsProvider() {
+		rg.w.Writefmtln("")
+		rg.w.Writefmtln("        private %s(string name, Input<string> id, %s? state = null, %s? options = null)",
+			className, rg.res.statet.name, optionsType)
+		rg.w.Writefmtln("            : base(\"%s\", name, state, MakeResourceOptions(options, id))", rg.res.info.Tok)
+		rg.w.Writefmtln("        {")
+		rg.w.Writefmtln("        }")
+	}
+
+	// Write the method that will calculate the resource options.
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("        private static %[1]s MakeResourceOptions(%[1]s? options, Input<string>? id)", optionsType)
+	rg.w.Writefmtln("        {")
+	rg.w.Writefmtln("            var defaultOptions = new %s", optionsType)
+	rg.w.Writefmtln("            {")
+	rg.w.Writefmtln("                Id = id,")
+	rg.w.Writefmt("                Version = Utilities.Version")
+
+	switch len(rg.res.info.Aliases) {
+	case 0:
+		rg.w.Writefmtln("")
+	case 1:
+		rg.w.Writefmtln(",")
```

Comment:
see abvoe.  if you just give everythign a trailing comma, you won't need this.  it's very normal for initializers to use that pattern.